lib/gruf/controllers/autoloader.rb

```diff
@@ -49,11 +51,19 @@ def initialize(path:, reloading: nil, tag: nil)
       def reload
         return unless @reloading_enabled
 
-        reload_mutex do
+        reload_lock.with_write_lock do
           @loader.reload
         end
       end
 
+      def with_fresh_controller(controller_name)
+        Gruf::Autoloaders.reload
```

Comment:
IMO, we should probably have this read:

```ruby
def with_fresh_controller(controller_name)
  return yield(controller_name.constantize) unless @reloading_enabled

  ::Gruf::Autoloaders.reload
  reload_lock.with_read_lock do
    yield(controller_name.constantize)
  end
end
```

I realize this does the extra check here, which is done in Gruf::Controllers::Autoloader, but the reason is two-fold:
1) We avoid calling the autoloader entirely here, which saves some CPU clicks
2) We avoid the RW lock entirely, which has a significant performance impact:

I ran a quick script like so:

```ruby
require 'concurrent/atomic/read_write_lock'
rw = Concurrent::ReadWriteLock.new
cycles = 10_000
Benchmark.bm do |x|
  x.report("with_read_lock") do
    cycles.times do
      rw.with_read_lock do
        Math.sqrt(4)
      end
    end
  end
  x.report("w/o_read_lock") do
    cycles.times do
      Math.sqrt(4)
    end
  end
end
```

This produced results:

```
       user     system      total        real
with_read_lock  0.013227   0.000035   0.013262 (  0.013283)
w/o_read_lock  0.000483   0.000001   0.000484 (  0.000484)
```

This is statistically significant  enough for us to want to ensure that execution path does not run - so we should bail early in the method.

Reply:
Wow you're right - that's an order-of-magnitude difference!

Taking it a bit further - should we also skip `.name.constantize` when there is no reloading? Constantize is not a free operation, either, and if classes don't change, there's no reason to re-resolve the controller on every call.

Reply:
IIRC the `.constantize` there is necessary to ensure that Zeitwerk picks it up from an autoloading perspective. We should keep that there.